hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/WriteHandleMetadataUtils.java

```diff
@@ -0,0 +1,226 @@
+   * reading the file slice without the new log files and comparing it by reading file slice with the
+   * new log files written by the handle.
+   *
+   * @param partitionPath             Partition path
+   * @param fileId                    Corresponding file id
+   * @param fileSliceOpt              File slice
+   * @param newLogFiles               New log files
+   * @param status                    Write status corresponding to the last log file written by the handle
+   * @param hoodieTable               Hoodie Table
+   * @param taskContextSupplier       Task context supplier
+   * @param secondaryIndexDefns       Definitions for secondary index which need to be updated
+   * @param config                    Write config
+   * @param instantTime               Instant time of the commit
+   * @param writeSchemaWithMetaFields Write schema with metadata fields
+   */
+  static void trackMetadataIndexStatsForStreamingMetadataWrites(String partitionPath, String fileId, Option<FileSlice> fileSliceOpt, List<String> newLogFiles, WriteStatus status,
+                                                                HoodieTable hoodieTable, TaskContextSupplier taskContextSupplier, List<Pair<String, HoodieIndexDefinition>> secondaryIndexDefns,
+                                                                HoodieWriteConfig config, String instantTime, Schema writeSchemaWithMetaFields) {
+    // TODO: @see <a href="https://issues.apache.org/jira/browse/HUDI-9533">HUDI-9533</a> Optimise the computation for multiple secondary indexes
+    HoodieEngineContext engineContext = new HoodieLocalEngineContext(hoodieTable.getStorageConf(), taskContextSupplier);
+    HoodieReaderContext readerContext = engineContext.getReaderContextFactory(hoodieTable.getMetaClient()).getContext();
+
+    // For Append handle, we need to merge the records written by new log files with the existing file slice to check
+    // the corresponding updates for secondary index. It is possible the records written in the new log files are ignored
+    // by record merger.
+    secondaryIndexDefns.forEach(secondaryIndexDefnPair -> {
+      // fetch primary key -> secondary index for prev file slice.
+      Map<String, String> recordKeyToSecondaryKeyForPreviousFileSlice = fileSliceOpt.map(fileSlice -> {
+        try {
+          return SecondaryIndexRecordGenerationUtils.getRecordKeyToSecondaryKey(hoodieTable.getMetaClient(), readerContext, fileSlice, writeSchemaWithMetaFields,
+              secondaryIndexDefnPair.getValue(), instantTime, config.getProps(), false);
+        } catch (IOException e) {
+          throw new HoodieIOException("Failed to generate secondary index stats ", e);
+        }
+      }).orElse(Collections.emptyMap());
+
+      // fetch primary key -> secondary index for latest file slice including inflight.
+      FileSlice latestIncludingInflight = fileSliceOpt.orElse(new FileSlice(new HoodieFileGroupId(partitionPath, fileId), instantTime));
```

Comment:
shouldn't we always compute the newLogFiles for the incremental changes? why still reading the previous file slice here? The result maybe be correct, but the reduadant read IO/compasiton cost is huge.

Reply:
We need to merge the new log files with the previous file slice to know if the secondary key has actually changed or not. The final determination can only happen by reading the file slice with and without log files using the merge logic.